File: python/basics/linkedlist/328_odd_even_linked_list.py
'''
328_odd_even_linked_list.py
get the even nodes.. 

result should have the odd nodes.. then followed by the even nodes
even_head_node

for each node...
    get the even
    get the odd.
    continue
'''

from linkedlist import ListNode

def odd_even_linked_list(head: ListNode) -> ListNode:
    if not head:
        return head
    node = head
    even_head = ListNode(0)
    node_even = even_head
    prev = None
    while node:
        prev = node
        even = node.next
        node_even.next = even
        node_even = even
        odd = None
        if even:
            odd = even.next
            even.next = None
        #odd
        node.next = odd
        node = odd
    prev.next = even_head.next
    return head


#head = [1,2,3,4,5]
#Output: [1,3,5,2,4]
A = ListNode(1)
B = ListNode(2)
C = ListNode(3)
D = ListNode(4)
E = ListNode(5)

A.next = B
B.next = C
C.next = D
D.next = E
print(A.traverse())
result = odd_even_linked_list(A)
print(result.traverse())


#head = [2,1,3,5,6,4,7]
#Output: [2,3,6,7,1,5,4]

A = ListNode(2)
B = ListNode(1)
C = ListNode(3)
D = ListNode(5)
E = ListNode(6)
F = ListNode(4)
G = ListNode(7)

A.next = B
B.next = C
C.next = D
D.next = E
E.next = F
F.next = G

print(A.traverse())
result = odd_even_linked_list(A)
print(result.traverse())

File: problem_solving/linkedlist/cpp/328_odd_even_linked_list.cpp
/*
https://leetcode.com/problems/odd-even-linked-list/
328_odd_even_linked_list.py
get the even nodes.. 

result should have the odd nodes.. then followed by the even nodes
even_head_node

for each node...
    get the even
    get the odd.
    continue
*/

#include<iostream>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
using namespace std;
#include<algorithm>
#include<stack>


ListNode* odd_even_linked_list(ListNode* head) {
    ListNode* node = head;
    ListNode* odd = new ListNode(0);
    ListNode* even = new ListNode(0);
    auto odd_node = odd;
    auto even_node = even;
    while(node) {
        odd_node->next = node;
        odd_node = odd_node->next;
        node = node->next;

        if(node) {
            even_node->next = node;
            even_node = even_node->next;
            node = node->next;
        }
    }
    even_node->next = nullptr;
    odd_node->next = even->next;
    return odd->next;
}


int main() {
    //#head = [1,2,3,4,5]
    //#Output: [1,3,5,2,4]
    ListNode* A = new ListNode(1);
    ListNode* B = new ListNode(2);
    ListNode* C = new ListNode(3);
    ListNode* D = new ListNode(4);
    ListNode* E = new ListNode(5);
    
    A->next = B;
    B->next = C;
    C->next = D;
    D->next = E;
    print_linked_list(A);
    auto result = odd_even_linked_list(A);
    print_linked_list(result);
    
    
    //#head = [2,1,3,5,6,4,7]
    //#Output: [2,3,6,7,1,5,4]
    
    A = new ListNode(2);
    B = new ListNode(1);
    C = new ListNode(3);
    D = new ListNode(5);
    E = new ListNode(6);
    ListNode* F = new ListNode(4);
    ListNode* G = new ListNode(7);
    
    A->next = B;
    B->next = C;
    C->next = D;
    D->next = E;
    E->next = F;
    F->next = G;
    
    print_linked_list(A);
    result = odd_even_linked_list(A);
    print_linked_list(result);
}
